internal/host/hostcommands/instruction_manager.go

```diff
@@ -178,10 +190,90 @@ func (i *InstructionManager) GetNextSteps(ctx context.Context, host *models.Host
 	} else {
 		returnSteps.NextInstructionSeconds = defaultNextInstructionInSec
 	}
+
+	// If the agent isn't compatible with the service and the host is in a state that allows to
+	// upgrade the agent, then replace all the calculated steps with a single step to upgrade
+	// the agent:
+	if i.isAgentUpgradeAllowed(host) && !i.isAgentCompatible(host) {
+		log.WithFields(logrus.Fields{
+			"expected_image": i.config.AgentImage,
+			"actual_image":   host.DiscoveryAgentVersion,
+		}).Info(
+			"Agent image ins't compatible with the service, and it can be upgraded, " +
+				"will replace all the calculated steps with a single step to " +
+				"upgrade the image",
+		)
+		var err error
+		returnSteps.Instructions, err = i.upgradeAgentCmd.GetSteps(ctx, host)
+		if err != nil {
+			return models.Steps{}, err
+		}
+		returnSteps.PostStepAction = swag.String(models.StepsPostStepActionContinue)
+		returnSteps.NextInstructionSeconds = defaultNextInstructionInSec
+		events.SendUpgradeAgentStartedEvent(
+			ctx,
+			i.eventsHandler,
+			*host.ID,
+			hostutil.GetHostnameForMsg(host),
+			host.InfraEnvID,
+			host.ClusterID,
+			i.config.AgentImage,
+		)
+	}
+
 	logSteps(returnSteps, InfraEnvID, hostID, log)
 	return returnSteps, nil
 }
 
+// isAgentCompatible checks if the agent running in the host is compatible with what the service
+// expects.
+func (i *InstructionManager) isAgentCompatible(host *models.Host) bool {
+	// Note that ideally we should retrieve the details of the validation from the host, but
+	// that is a JSON document, and the data types used to define them are part of the 'host'
+	// package, which we can't import here because it would introduce a circular dependency.
+	// Instead of that we are performing the validation again. In this case it isn't a problem
+	// because the logic of the validation is very simple.
+	return common.IsAgentCompatible(i.config.AgentImage, host.DiscoveryAgentVersion)
+}
+
+// isAgentUpgradeAllowed checks if the current state of the host allows the agent to be upgraded.
+// For example, it is not allowed to upgrade the agent when the installation of the cluster is in
+// progress.
+func (i *InstructionManager) isAgentUpgradeAllowed(host *models.Host) bool {
+	if !i.config.EnableUpgradeAgent {
+		return false
+	}
+	if i.isStepDisabled(models.StepTypeUpgradeAgent) {
+		return false
+	}
+	switch *host.Status {
+	case models.HostStatusAddedToExistingCluster,
```

Comment:
i think that relevant statuses are:
models.HostStatusInsufficient,
models.HostStatusKnown,
not sure regarding late binding but i assume:
models.HostStatusBinding

Maybe i missed something but we should not allow so many statuses for sure
@avishayt ?

Reply:
I updated the list of states to reflect what we discussed and reviewed in the enhancement document: https://github.com/openshift/assisted-service/blob/master/docs/enhancements/agent-upgrade.md#implementation-detailsnotesconstraints . @eranco74, @tsorya I think it addresses your concerns.

Note that the states where the agent is not running aren't really relevant, because this code only runs when a request is received from a running agent. Anyhow I think that should be addressed now as well.